Spatial_Engine: Align centroid with origin when creating profiles

IsakNaslundBh opened the pull request:
<!-- PLEASE ENSURE YOU REVIEW THE CONTENT OF EACH PR CAREFULLY, INCLUDING SUBSEQUENT COMMENTS BY YOURSELF OR OTHERS. -->
<!-- IN PARTICULAR PLEASE ENSURE THAT SENSITIVE OR INAPPROPRIATE INFORMATION IS NOT UPLOADED -->

### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Closes #2148 

<!-- Add short description of what has been fixed -->

Making sure all non-symmetrical profiles have their centroid of the curves aligned with the origin
### Test files
<!-- Link to test files to validate the proposed changes -->

https://burohappold.sharepoint.com/:f:/s/BHoM/EvwQushTTIdBnbiI0b_XPD0BtYWuomLySsxlJY6Yb6W7IQ?e=OFgPjD

### Changelog
<!-- Text to go into changelog if applicable -->
<!-- Please see https://github.com/BHoM/documentation/wiki/changelog for guidelines -->


### Additional comments
<!-- As required -->

pawelbaran commented on Spatial_Engine/Create/ShapeProfiles/FabricatedBoxProfile.cs:
```diff
@@ -70,6 +70,8 @@ public static FabricatedBoxProfile FabricatedBoxProfile(double height, double wi
             }
 
             List<ICurve> curves = FabricatedBoxProfileCurves(width, height, webThickness, topFlangeThickness, botFlangeThickness, weldSize);
+
+
```
But why 🐒

IsakNaslundBh replied:
because!

FraserGreenroyd replied:
Well don't

pawelbaran commented on Spatial_Engine/Create/ShapeProfiles/FreeFormProfile.cs:
```diff
@@ -136,13 +138,29 @@ public static FreeFormProfile FreeFormProfile(IEnumerable<ICurve> edges)
-
             }
 
+            if (!couldCentre)
```
Couldn't that sit in the `catch` from line 151? `couldCentre` could possibly be avoided then.

IsakNaslundBh replied:
was in case it is a nurbs or ellipse. Defined the variable outside the two if statements. So if it contains nurbs or ellipses, or crashes while trying, this will be false and raise the message, if that makes sense?

pawelbaran replied:
Ah ok, makes perfect sense after a second look.

pawelbaran commented on Spatial_Engine/Create/ShapeProfiles/TSectionProfile.cs:
```diff
@@ -112,6 +112,11 @@ private static List<ICurve> TeeProfileCurves(double tft, double tfw, double wt,
+            Point centroid = perimeter.IJoin().Centroid();
+            Vector tranlation = Point.Origin - centroid;
+
+            perimeter = perimeter.Select(x => x.ITranslate(tranlation)).ToList();
```
You could return in this line already (same in other files)

IsakNaslundBh replied:
ah, true, will fix that

pawelbaran reviewed (approved):
LGTM, works fine 👍 I have left a few purely aesthetic comments, please feel free to ignore 😉

IsakNaslundBh commented:
Changed to simply returning the linq query, as you suggested, and moved all logic to the curve methods. After thinking about it for two more seconds, I realised that we are mirroring about the global origin anyway, so if the curves are cetered around that already, mirroring before/after will make no difference! 🛩️

pawelbaran commented on Spatial_Engine/Create/ShapeProfiles/AngleProfile.cs:
```diff
@@ -121,10 +117,9 @@ private static List<ICurve> AngleProfileCurves(double width, double depth, doubl
             perimeter.Add(new Line { Start = p, End = p = p - yAxis * (depth) });
             List<ICurve> translatedCurves = new List<ICurve>();
```
looks like this variable is not used anywhere

pawelbaran commented on Spatial_Engine/Create/ShapeProfiles/AngleProfile.cs:
```diff
@@ -121,10 +117,9 @@ private static List<ICurve> AngleProfileCurves(double width, double depth, doubl
-
-            return translatedCurves;
+            Point centroid = perimeter.IJoin().Centroid();
+            Vector tranlation = Point.Origin - centroid;
```
tranSlation

pawelbaran commented on Spatial_Engine/Create/ShapeProfiles/ChannelProfile.cs:
```diff
@@ -111,7 +107,9 @@ private static List<ICurve> ChannelProfileCurves(double height, double width, do
 
-            return edges;
+            Point centroid = edges.IJoin().Centroid();
+            Vector tranlation = Point.Origin - centroid;
```
tranSlation

pawelbaran commented on Spatial_Engine/Create/ShapeProfiles/FabricatedBoxProfile.cs:
```diff
@@ -114,9 +112,7 @@ private static List<ICurve> FabricatedBoxProfileCurves(double width, double heig
-            box = box.Select(x => x.ITranslate(tranlation)).ToList();
-
-            return box;
+            return box.Select(x => x.ITranslate(tranlation)).ToList();
```
tranSlation

pawelbaran commented on Spatial_Engine/Create/ShapeProfiles/GeneralisedFabricatedBoxProfile.cs:
```diff
@@ -122,9 +122,7 @@ private static List<ICurve> GeneralisedFabricatedBoxProfileCurves(double height,
-            group = group.Select(x => x.ITranslate(tranlation)).ToList();
-
-            return group;
+            return group.Select(x => x.ITranslate(tranlation)).ToList();
```
tranSlation

pawelbaran commented on Spatial_Engine/Create/ShapeProfiles/GeneralisedTSectionProfile.cs:
```diff
@@ -116,9 +116,7 @@ private static List<ICurve> GeneralisedTeeProfileCurves(double height, double we
+            return perimeter.Select(x => x.ITranslate(tranlation)).ToList();
         }
 
         /***************************************************/
```
tranSlation

pawelbaran commented on Spatial_Engine/Create/ShapeProfiles/KiteProfile.cs:
```diff
@@ -115,9 +115,7 @@ private static List<ICurve> KiteProfileCurves(double width1, double angle1, doub
-            group = group.Select(x => x.ITranslate(tranlation)).ToList();
-
-            return group;
+            return group.Select(x => x.ITranslate(tranlation)).ToList();
```
tranSlation

pawelbaran commented on Spatial_Engine/Create/ShapeProfiles/TSectionProfile.cs:
```diff
@@ -115,9 +115,7 @@ private static List<ICurve> TeeProfileCurves(double tft, double tfw, double wt,
-            perimeter = perimeter.Select(x => x.ITranslate(tranlation)).ToList();
-
-            return perimeter;
+            return perimeter.Select(x => x.ITranslate(tranlation)).ToList();
```
tranSlation

pawelbaran reviewed (changes requested):
I also gave it a 2nd look and ofc found something to complain about 🙈

FraserGreenroyd commented:
@BHoMBot check compliance

FraserGreenroyd commented:
@BHoMBot check documentation compliance

FraserGreenroyd commented:
@BHoMBot check documentation compliance

FraserGreenroyd commented:
@BHoMBot check documentation compliance

FraserGreenroyd commented:
@BHoMBot check documentation compliance

FraserGreenroyd commented:
@BHoMBot check documentation compliance

FraserGreenroyd commented:
@BHoMBot check project compliance

FraserGreenroyd commented:
@BHoMBot check project compliance

FraserGreenroyd commented:
@BHoMBot check project compliance

FraserGreenroyd commented:
@BHoMBot check project compliance

pawelbaran reviewed (approved):
LGTM

IsakNaslundBh commented:
/azp run BHoM_Engine.CheckInstaller

FraserGreenroyd commented:
@BHoMBot check code compliance

FraserGreenroyd commented:
@BHoMBot check code compliance

FraserGreenroyd commented:
@BHoMBot check code compliance
